src/Compilers/CSharp/Portable/Symbols/Synthesized/GeneratedNames.cs

```diff
@@ -54,17 +54,34 @@ internal static string MakeLambdaDisplayClassName(int methodOrdinal, int generat
             return MakeMethodScopedSynthesizedName(GeneratedNameKind.LambdaDisplayClass, methodOrdinal, generation, suffix: "DisplayClass", entityOrdinal: closureOrdinal, entityGeneration: closureGeneration, isTypeName: true);
         }
 
-        internal static string MakeAnonymousTypeTemplateName(int index, int submissionSlotIndex, string moduleId)
+        private static string MakeTopLevelTypeName(string typePostfix, int index, int submissionSlotIndex, string moduleId)
         {
-            var name = "<" + moduleId + ">f__AnonymousType" + StringExtensions.GetNumeral(index);
-            if (submissionSlotIndex >= 0)
+            var name = "<" + moduleId + typePostfix + StringExtensions.GetNumeral(index);
+            if (submissionSlotIndex > 0)
             {
                 name += "#" + StringExtensions.GetNumeral(submissionSlotIndex);
             }
-
             return name;
         }
 
+        internal static string MakeMethodGroupConversionCacheFrameName(int index, int submissionSlotIndex, string moduleId)
+        {
+            Debug.Assert((char)GeneratedNameKind.MethodGroupConversionCacheFrame == 'q');
+
+            return MakeTopLevelTypeName(">q__", index, submissionSlotIndex, moduleId);
+        }
+
+        internal static string MakeMethodGroupConversionCacheDelegateFieldName(string methodName)
+        {
+            Debug.Assert((char)GeneratedNameKind.MethodGroupConversionCacheDelegateField == 'r');
+            return "<" + methodName + ">r";
+        }
+
+        internal static string MakeAnonymousTypeTemplateName(int index, int submissionSlotIndex, string moduleId)
+        {
+            return MakeTopLevelTypeName(">f__AnonymousType", index, submissionSlotIndex, moduleId);
```

Comment:
Not your change, but shouldn't we be asserting that some `GeneratedNameKind` is equal to `'f'`?